src/messages/embed/Embed.ts

```diff
@@ -0,0 +1,304 @@
+	 * The embed image data.
+	 */
+	public image?: APIEmbedImage;
+
+	/**
+	 * Received video data.
+	 */
+	public video?: APIEmbedVideo;
+
+	/**
+	 * The embed author data.
+	 */
+	public author?: APIEmbedAuthor;
+
+	/**
+	 * Received data about the embed provider.
+	 */
+	public provider?: APIEmbedProvider;
+
+	/**
+	 * The embed footer data.
+	 */
+	public footer?: APIEmbedFooter;
+
+	public constructor(data: APIEmbed = {}) {
+		this.title = data.title;
+		this.description = data.description;
+		this.url = data.url;
+		this.color = data.color;
+		this.thumbnail = data.thumbnail;
+		this.image = data.image;
+		this.video = data.video;
+		this.author = data.author;
+		this.provider = data.provider;
+		this.footer = data.footer;
+		this.fields = data.fields ?? [];
+
+		if (data.timestamp) this.timestamp = new Date(data.timestamp).toISOString();
+	}
+
+	/**
+	 * The accumulated length for the embed title, description, fields, footer text, and author name.
+	 */
+	public get length(): number {
+		return (
+			(this.title?.length ?? 0) +
+			(this.description?.length ?? 0) +
+			(this.fields.length >= 1
+				? this.fields.reduce((prev, curr) => prev + curr.name.length + curr.value.length, 0)
+				: 0) +
+			(this.footer?.text.length ?? 0) +
+			(this.author?.name?.length ?? 0)
+		);
+	}
+
+	/**
+	 * Adds a field to the embed (max 25).
+	 * @param name The name of this field.
+	 * @param value The value of this field.
+	 * @param inline If this field will be displayed inline.
+	 */
+	public addField(name: string, value: string, inline = false): this {
+		return this.addFields({ name, value, inline });
+	}
+
+	/**
+	 * Adds fields to the embed (max 25).
+	 * @param fields The fields to add.
+	 */
+	public addFields(...fields: APIEmbedField[]): this {
+		// Data assertions
+		ow(fields, embedFieldsArrayPredicate);
+
+		// Ensure adding this field won't exceed the 25 field limit
+		validateFieldLength(this.fields, fields.length);
+
+		this.fields.push(...Embed.normalizeFields(...fields));
+		return this;
+	}
+
+	/**
+	 * Removes, replaces, and inserts fields in the embed (max 25).
+	 * @param index The index to start at.
+	 * @param deleteCount The number of fields to remove.
+	 * @param fields The replacing field objects.
+	 */
+	public spliceFields(index: number, deleteCount: number, ...fields: APIEmbedField[]): this {
+		// Data assertions
+		ow(fields, 'fields', embedFieldsArrayPredicate);
+
+		// Ensure adding this field won't exceed the 25 field limit
+		validateFieldLength(this.fields, fields.length - deleteCount);
+
+		this.fields.splice(index, deleteCount, ...Embed.normalizeFields(...fields));
+		return this;
+	}
+
+	/**
+	 * Sets the author of this embed.
+	 * @param name The name of the author.
+	 * @param options The URL options of the author.
+	 * @param options.iconURL The icon URL of the author.
+	 * @param options.url The URL of the author.
+	 */
+	public setAuthor(name: string | null, { iconURL, url }: AuthorURLOptions = {}): this {
```

Comment:
Personally I dislike the "required parameter goes alone and optionals in an object". It's also why slash commands is all builders... I'd rather this become either a builder or an object all the time, but I'll boop @discordjs/the-big-3 for their input